perfkitbenchmarker/linux_virtual_machine.py

```diff
@@ -974,3 +976,158 @@ def MoveFile(self, target, source_path, remote_path=''):
     target.ContainerCopy(file_name, remote_path)
+
+
+class JujuMixin(DebianMixin):
+  """
+  Class to allow running juju-deployed workloads
+
+  Bootstraps a Juju environment using the manual provider:
+  https://jujucharms.com/docs/stable/config-manual
+  """
+
+  OS_TYPE = os_types.JUJU
+
+  isController = False
+
+  """
+  A reference to the juju controller, useful when operations occur against
+  a unit's VM but need to be preformed from the controller.
+  """
+  controller = None
+
+  units = []
+
+  environments_yaml = """
+  default: perfkit
+
+  environments:
+      perfkit:
+          type: manual
+          bootstrap-host: {0}
+  """
+
+  def _Bootstrap(self):
+    resp, _ = self.RemoteHostCommand("juju bootstrap")
+
+  def JujuAddMachine(self, ip):
+    resp, _ = self.RemoteHostCommand("juju add-machine ssh:%s" % ip)
+
+  def JujuConfigureEnvironment(self):
+    if self.isController:
+      resp, _ = self.RemoteHostCommand("mkdir -p ~/.juju")
+      f = tempfile.NamedTemporaryFile(delete=False)
+      f.write(self.environments_yaml.format(self.internal_ip))
+      f.close()
+      self.RemoteCopy(f.name, "~/.juju/environments.yaml")
+      os.unlink(f.name)
+
+  def JujuEnvironment(self):
+    output, _ = self.RemoteHostCommand('juju switch')
+    return output.strip()
+
+  def JujuRun(self, cmd):
+    output, _ = self.RemoteHostCommand(cmd)
+    return output.strip()
+
+  def JujuStatus(self):
+    """
+    Return the status of the Juju environment.
+    """
+    output, _ = self.RemoteHostCommand('juju status --format=json')
+    return output.strip()
+
+  def JujuVersion(self):
+    output, _ = self.RemoteHostCommand('juju version')
+    return output.strip()
+
+  def JujuSet(self, service, params=[]):
+    ' '.join(params)
+    output, _ = self.RemoteHostCommand(
+        'juju set %s %s' % (service, ' '.join(params)))
+    return output.strip()
+
+  def JujuWait(self, timeout=900):
+    """
+    Wait until all services are deployed, related or idle
+    """
+
+    start = time.time()
+    while True:
+      ready = True
+      status = yaml.load(self.JujuStatus())
+      for service in status['services']:
+        ss = status['services'][service]['service-status']['current']
+
+        # Accept blocked because the service may be waiting on relation
+        if ss not in ['active', 'unknown']:
+          ready = False
+
+        if ss in ['error']:
+          # The service has failed to deploy.
+          debuglog = self.JujuRun('juju debug-log --limit 200')
+          logging.warn(debuglog)
+          return False
+
+        for unit in status['services'][service]['units']:
+          unit_data = status['services'][service]['units'][unit]
+          ag = unit_data['agent-state']
+          if ready and ag != 'started':
+            ready = False
+
+          ws = unit_data['workload-status']['current']
+          if ready and ws not in ['active', 'unknown']:
+            ready = False
+
+      if start + timeout < time.time():
```

Comment:
I don't think this does what you intended it to do - it looks like it will always set ready to false after the timeout is reached instead of returning/timing out